freqtrade/freqtradebot.py

```diff
@@ -916,17 +913,7 @@ def handle_timedout_limit_buy(self, trade: Trade, order: Dict) -> bool:
         # we need to fall back to the values from order if corder does not contain these keys.
         trade.amount = order['amount'] - corder.get('remaining', order['remaining'])
         trade.stake_amount = trade.amount * trade.open_rate
-        # verify if fees were taken from amount to avoid problems during selling
-        try:
-            new_amount = self.get_real_amount(trade, corder if 'fee' in corder else order,
-                                              trade.amount)
-            if not isclose(order['amount'], new_amount, abs_tol=constants.MATH_CLOSE_PREC):
-                trade.amount = new_amount
-                # Fee was applied, so set to 0
-                trade.fee_open = 0
-                trade.recalc_open_trade_price()
-        except DependencyException as e:
-            logger.warning("Could not update trade amount: %s", e)
+        self.update_trade_state(trade, corder if 'fee' in corder else order, trade.amount)
```

Comment:
Hmm, I'm thinking of this expression:
```
corder if 'fee' in corder else order
```

I see, it's from the original code, but do you remember/know why is it in this (quite) tricky way?

Reply:
we're trying to use the fees from the latest possible point - which ideally is from the `cancel_order()` call - available as `corder`.

The result of cancel_order is not really consistent between exchanges - some return the same as `fetch_order()` (but with the state right after the order was cancelled) - while others return a simple string in the form of "success" (FTX - i'm looking at you).

For that reason, we must use the result from `fetch_order()` (available as order) - verify it's a dict (done a few lines above) - and then always use corder (or fallback to order) for everything we use from corder.

#3146 will enhance drive this logic further - but keeping these 2 PR's seperate is better as the 2 address different problems (this one code duplication and potential timing hickups  - the other one inconsistent content in the order and cancel_order results).

Reply:
> The result of cancel_order is not really consistent between exchanges

Yeah, I remember this fact and that fix you made...

> and then always use corder (or fallback to order) for everything we use from corder

yes, this (as written here) is clean, I wonder the manner the check is performed:
```
corder if 'fee' in corder else order
```
-- why 'fee' is used here, why not some other element?

Reply:
well because the most releant place is the call to get_real_amount ... which does heavily rely on fee to be available to work correctly.

Also, fee is the element that's the "least secure" - it's missing in most exchanges ... (but not in all)